tests/functional/u/use/used_before_assignment_except_handler_for_try_with_return.py

```diff
@@ -49,3 +49,66 @@ def func_ok3(var):
     except ZeroDivisionError:
         msg = "Division by 0"
     print(msg)
+
+
+def func_ok4(var):
+    """Define "msg" with a chained assignment."""
+    try:
+        return 1 / var.some_other_func()
+    except AttributeError:
+        msg2 = msg = "Division by 0"
+        print(msg2)
+    except ZeroDivisionError:
+        msg = "Division by 0"
+    print(msg)
+
+
+def func_ok5(var):
+    """Define 'msg' via unpacked iterable."""
+    try:
+        return 1 / var.some_other_func()
+    except AttributeError:
+        msg, msg2 = ["Division by 0", "Division by 0"]
+        print(msg2)
+    except ZeroDivisionError:
+        msg = "Division by 0"
+    print(msg)
+
+
+def func_invalid1(var):
+    """'msg' is not defined in one handler."""
+    try:
+        return 1 / var.some_other_func()
+    except AttributeError:
+        pass
+    except ZeroDivisionError:
+        msg = "Division by 0"
+    print(msg)  # [used-before-assignment]
+
+
+def func_invalid2(var):
+    """'msg' is not defined in one handler."""
+    try:
+        return 1 / var.some_other_func()
+    except AttributeError:
+        msg: str
+    except ZeroDivisionError:
+        msg = "Division by 0"
+    print(msg)  # [used-before-assignment]
+
+
+def func_invalid3(var):
```

Comment:
These tests are great! Just one remark: `func_invalid3` is not actually testing anything new compared to `func_invalid2`. I think we want to test the assignment in `if` statements here, which would be best tested by removing the `AttributeError` except handler and checking if not message is being raised. Currently this just tests that `msg: str` works as expected, if the handling of the `if` statement is incorrect we wouldn't notice.

Reply:
Actually, I think it does test something different. The first implementation didn't catch this case (by @areveny), and we discussed it [here](https://github.com/PyCQA/pylint/pull/5764#discussion_r798116392).

Should I flesh out the docstring to clarify?

Reply:
Hmm could it be that something changed in the implementation that made that test different? 

```python
def func_invalid3(var):
    """'msg' is not defined in one handler."""
    err_message = False
    try:
        return 1 / var.some_other_func()
    except ZeroDivisionError:
        if err_message:
            msg = "Division by 0"
        else:
            msg = None
    print(msg)
```

This shouldn't raise right?

Reply:
Sorry, I don't think I'm following you very well. 😅 . I thought you were proposing I take `func_invalid3` and rewrite it like you just did in the comment directly above this one. But I'm saying that `func_invalid3` as written catches a false negative that this PR didn't address until @areveny pointed it out and I fixed it (to stop assuming direct parentage), so I think the test has value as written. But I think I may be missing your point still, perhaps? Would you like to me _add_ this as another test, so it becomes `func_ok6`?

Reply:
Yeah I'm not making many sense. Let me rephrase:
I don't really understand why the third function needs to be tested in combination with the `msg: str` statement. Why would the behaviour of the message change when there is a combination of `msg: str` and `ifs` in two expect handlers. Imo it makes more sense to test those separately.
However, you have discussed this a lot with @areveny so perhaps there is something difficult going on when both of those patterns are present and we should test them together? Then perhaps the docstring should be added.
While trying to find out if I could find this relatedness between `msg: str` and the `ifs` I stumbled upon the example I gave above. I think the current version raises on this code but it shouldn't, right? That is a separate issue from the one with which I started this comment thread, but it might be related.

Sorry for being so unclear 😅 Long day..

Reply:
No problem, I can explain.

> However, you have discussed this a lot with @areveny so perhaps there is something difficult going on when both of those patterns are present and we should test them together?

In a word, yup.

I can change `msg: str` to `pass` if that would be more clear that it doesn't have to do with type annotations. However, having two handlers where the one with the name is nested under if or with is the essence of the test case. All of the tests for correct positives require multiple except handlers. This one in involves nesting under an if or with just to catch the faulty assumption the code first made about direct parentage. 

"Round 1" of most of my control flow PRs adding methods like `_uncertain_nodes_in...` just assumed direct parentage of the node being assigned by its containing try/except. After you pointed me to `utils.get_node_first_ancestor_of_type` I did #5582 to stop assuming this and catch more false negatives, but that PR didn't touch this method. (I guess I was trying to take small bites! 😅 ) @areveny spotted the false negative, and I decided it should be done in this PR. So if you check out the first commit on the branch, you'll see it fails the test case as written.

***
And then you found a false positive for me to investigate! I'll see what's up with that.

Reply:
> I can change `msg: str` to `pass` if that would be more clear that it doesn't have to do with type annotations. However, having two handlers where the one with the name is nested under if or with is the essence of the test case. All of the tests for correct positives require multiple except handlers. This one in involves nesting under an if or with just to catch the faulty assumption the code first made about direct parentage.

But if there was only a single expect handler with nested if and the message isn't raised, won't the functionality of nested ifs be tested as well? If it does not work (ie., assignment in nested ifs is not recognised) the message would be raised incorrectly. 

> And then you found a false positive for me to investigate! I'll see what's up with that.

👍

Reply:
> But if there was only a single expect handler with nested if and the message isn't raised, won't the functionality of nested ifs be tested as well? If it does not work (ie., assignment in nested ifs is not recognised) the message would be raised incorrectly.

Ah, no, I don't think testing a correct positive and correct negative will necessarily go down the same code path. You're talking about testing a correct negative (which is good to test!), but that will short-circuit a lot earlier (see `"# If the other node is in the same scope as this node, assume it executes"`), which is why it was probably passing.